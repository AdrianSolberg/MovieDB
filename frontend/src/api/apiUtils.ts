import { Review } from "@/types/__generated__/types";

type ResponseReview = {
    __typename?: "Review";
    _id: string;
    username: string;
    rating: number;
    comment: string;
    date: Date;
};

/**
 *  Make sure that the date-field is correctly initialized as a Date-object
 *  when getting reviews as a response from a query.
 */
export function parseReviewResponse(inputReviews: ResponseReview[]): Review[] {
    return inputReviews.map((review) => ({
        ...review,
        date: new Date(review.date),
    })) as Review[];
}
