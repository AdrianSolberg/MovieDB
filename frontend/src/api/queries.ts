import { gql } from "@/types/__generated__";

export const GET_MOVIES = gql(`
    query GetMovies(
        $skip: Int
        $limit: Int
        $filters: FiltersInput
        $sortOption: SortingType
        $search: String
    ) {
        movies(
            skip: $skip
            limit: $limit
            filters: $filters
            sortOption: $sortOption
            search: $search
        ) {
            _id
            title
            vote_average
            release_date
            runtime
            poster_path
        }
    }
`);

export const GET_MOVIE = gql(`
    query GetMovie($movieId: Int!) {
        movie(id: $movieId) {
            _id
            title
            vote_average
            vote_count
            status
            release_date
            revenue
            runtime
            backdrop_path
            budget
            homepage
            imdb_id
            original_language
            original_title
            overview
            popularity
            poster_path
            tagline
            genres
            production_companies
            production_countries
            spoken_languages
            keywords
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);

export const GET_FILTERS = gql(`
    query GetFilters {
        filters {
            Genre
            Rating
            Decade
            Status
            Runtime
        }
    }
`);

export const GET_LATEST_REVIEWS = gql(`
    query GetLatestReviews(
        $skip: Int,
        $limit: Int
    ) {
        latestReviews(
            skip: $skip,
            limit: $limit
        ) {
            _id
            movie {
                _id
                title
                poster_path
            }
            username
            rating
            comment
            date
        }
    }
`);

export const GET_USER_REVIEWS = gql(`
    query GetUserReviews(
        $username: String!,
        $skip: Int,
        $limit: Int
    ) {
        userReviews(
            username: $username,
            skip: $skip,
            limit: $limit
        ) {
            _id
            movie {
                _id
                title
                poster_path
            }
            rating
            comment
            date
        }
    }
`);

export const ADD_REVIEW = gql(`
    mutation AddReview(
        $movieId: Int!,
        $username: String!,
        $rating: Int!,
        $comment: String!
    ) {
        addReview(
            movie_id: $movieId,
            username: $username,
            rating: $rating,
            comment: $comment
        ) {
            _id
            vote_average
            vote_count
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);

export const DELETE_REVIEW = gql(`
    mutation DeleteReview($id: ID!) {
        deleteReview(_id: $id) {
            _id
            vote_average
            vote_count
            reviews {
                _id
                username
                rating
                comment
                date
            }
        }
    }
`);
