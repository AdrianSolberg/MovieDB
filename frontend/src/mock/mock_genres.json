[
    {
        "id": 0,
        "name": "Drama"
    },
    {
        "id": 1,
        "name": "Romance"
    },
    {
        "id": 2,
        "name": "Western"
    },
    {
        "id": 3,
        "name": "Animation"
    },
    {
        "id": 4,
        "name": "War"
    },
    {
        "id": 5,
        "name": "Science Fiction"
    },
    {
        "id": 6,
        "name": "Comedy"
    },
    {
        "id": 7,
        "name": "Fantasy"
    },
    {
        "id": 8,
        "name": "Adventure"
    },
    {
        "id": 9,
        "name": "Crime"
    },
    {
        "id": 10,
        "name": "Thriller"
    },
    {
        "id": 11,
        "name": "Family"
    },
    {
        "id": 12,
        "name": "Action"
    },
    {
        "id": 13,
        "name": "Mystery"
    }
]
