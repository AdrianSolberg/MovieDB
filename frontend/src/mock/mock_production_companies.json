[
    "Lynda Obst Productions",
    "Pascal Pictures",
    "Fairview Entertainment",
    "Silver Pictures",
    "Walt Disney Pictures",
    "Visiona Romantica",
    "DC Films",
    "Mandate International",
    "Patalex IV Productions Limited",
    "Amblin Entertainment",
    "Heyday Films",
    "Universal Pictures",
    "Bron Studios",
    "Juno Pix",
    "Color Force",
    "LStar Capital",
    "Miramax",
    "The Saul Zaentz Company",
    "Sikelia Productions",
    "Red Granite Pictures",
    "The Weinstein Company",
    "Knickerbocker Films",
    "Columbia Pictures",
    "Groucho II Film Partnership",
    "Syncopy",
    "P of A Productions Limited",
    "Zehnte Babelsberg Film",
    "The Linson Company",
    "Pixar",
    "Appian Way",
    "The Steve Tisch Company",
    "Genre Films",
    "Paramount",
    "Kennedy Miller Productions",
    "Lucasfilm Ltd.",
    "New Line Cinema",
    "Atman Entertainment",
    "Cecchi Gori Pictures",
    "Castle Rock Entertainment",
    "Isobel Griffiths",
    "Phoenix Pictures",
    "DC Entertainment",
    "Dune Entertainment",
    "20th Century Fox",
    "Marvel Studios",
    "Lightstorm Entertainment",
    "WingNut Films",
    "Fox 2000 Pictures",
    "RatPac Entertainment",
    "Lionsgate",
    "Warner Bros. Pictures",
    "Patalex III Productions Limited",
    "Jersey Films",
    "Joint Effort",
    "Village Roadshow Pictures",
    "EMJAG Productions",
    "The Donners' Company",
    "1492 Pictures",
    "A Band Apart",
    "DC Comics",
    "Atlas Entertainment",
    "Ingenious Media",
    "Legendary Pictures",
    "Taurus Film",
    "Wendy Finerman Productions",
    "Peninsula Films",
    "Lin Pictures",
    "Regency Enterprises",
    "Jerry Bruckheimer Films"
]
