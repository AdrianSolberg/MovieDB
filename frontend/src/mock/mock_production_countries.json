[
    {
        "id": 0,
        "name": "New Zealand"
    },
    {
        "id": 1,
        "name": "Canada"
    },
    {
        "id": 2,
        "name": "South Africa"
    },
    {
        "id": 3,
        "name": "United Kingdom"
    },
    {
        "id": 4,
        "name": "Germany"
    },
    {
        "id": 5,
        "name": "Australia"
    },
    {
        "id": 6,
        "name": "France"
    },
    {
        "id": 7,
        "name": "United States of America"
    }
]
