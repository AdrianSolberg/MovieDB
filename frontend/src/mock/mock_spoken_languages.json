[
    {
        "id": 0,
        "name": "Swahili"
    },
    {
        "id": 1,
        "name": "French"
    },
    {
        "id": 2,
        "name": "German"
    },
    {
        "id": 3,
        "name": "Persian"
    },
    {
        "id": 4,
        "name": "Italian"
    },
    {
        "id": 5,
        "name": "Japanese"
    },
    {
        "id": 6,
        "name": "Korean"
    },
    {
        "id": 7,
        "name": "Hindi"
    },
    {
        "id": 8,
        "name": "Urdu"
    },
    {
        "id": 9,
        "name": "English"
    },
    {
        "id": 10,
        "name": "Mandarin"
    },
    {
        "id": 11,
        "name": "Spanish"
    },
    {
        "id": 12,
        "name": "Russian"
    },
    {
        "id": 13,
        "name": "Xhosa"
    },
    {
        "id": 14,
        "name": "Romanian"
    },
    {
        "id": 15,
        "name": "Norwegian"
    },
    {
        "id": 16,
        "name": "Swedish"
    },
    {
        "id": 17,
        "name": "Arabic"
    }
]
