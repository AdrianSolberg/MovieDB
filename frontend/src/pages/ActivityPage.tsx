import ReviewCard from "@/components/ReviewCard";
import InfiniteScroll from "react-infinite-scroller";
import { useMemo, useState } from "react";
import { Link } from "react-router-dom";
import { useQuery } from "@apollo/client";
import { GET_LATEST_REVIEWS } from "../api/queries";
import { Review } from "@/types/__generated__/types";
import Loader from "../components/Loader";

const ActivityPage = () => {
    const [isMoreReviews, setIsMoreReviews] = useState<boolean>(false);
    const LIMIT = 20;

    const { data, loading, error, fetchMore } = useQuery(GET_LATEST_REVIEWS, {
        variables: {
            skip: 0,
            limit: LIMIT,
        },
        onCompleted: (data) => {
            if (data.latestReviews) {
                setIsMoreReviews(
                    data.latestReviews.length >= LIMIT &&
                        data.latestReviews.length % LIMIT === 0
                );
            }
        },
    });

    const latestReviews = useMemo(
        () => data?.latestReviews as Review[] | undefined,
        [data]
    );

    const handleLoadMore = () => {
        fetchMore({ variables: { skip: latestReviews?.length } }).then(
            (fetchMoreResult) => {
                setIsMoreReviews(
                    fetchMoreResult.data.latestReviews.length === LIMIT
                );
            }
        );
    };

    if (loading) {
        return (
            <section className="mt-6 w-dvw text-center">
                <Loader size="lg">
                    <p className="text-2xl">Loading...</p>
                </Loader>
            </section>
        );
    }

    if (error || !latestReviews) {
        return (
            <section className="mt-6 w-dvw text-center">
                <h1 className="text-2xl">Something went wrong!</h1>
                <p className="text-primary">Try to refresh</p>
            </section>
        );
    }

    if (latestReviews.length === 0) {
        return (
            <section className="mt-6 text-center">
                <h1 className="text-2xl">No reviews have been added yet</h1>
                <Link to="/" className="text-primary hover:underline">
                    Return to home page
                </Link>
            </section>
        );
    }

    return (
        <main id="main-content" className="mx-auto mt-8 max-w-6xl px-4 pb-4">
            <h1 className="mb-6 text-center text-3xl font-bold">
                Latest Activity
            </h1>
            <InfiniteScroll
                loadMore={handleLoadMore}
                hasMore={isMoreReviews}
                initialLoad={false}
                threshold={100}
                loader={
                    <Loader
                        key={-1}
                        aria-label="Loading more reviews..."
                        className="text-center"
                    >
                        <p className="text-2xl">Loading...</p>
                    </Loader>
                }
            >
                <ul className="space-y-6">
                    {latestReviews.map((review) => (
                        <li key={review._id}>
                            <ReviewCard review={review} />
                        </li>
                    ))}
                </ul>
            </InfiniteScroll>
        </main>
    );
};

export default ActivityPage;
