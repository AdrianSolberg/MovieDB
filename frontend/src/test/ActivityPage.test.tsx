import { GET_LATEST_REVIEWS } from "@/api/queries";
import { all_reviews } from "./mock/util";
import "@testing-library/jest-dom";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import ActivityPage from "@/pages/ActivityPage";
import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";

const mockLatestReviewsEmpty = [
    {
        request: {
            query: GET_LATEST_REVIEWS,
            variables: {
                skip: 0,
                limit: 20,
            },
        },
        result: {
            data: {
                latestReviews: [],
            },
        },
    },
];

const mockLatestReviews = [
    {
        request: {
            query: GET_LATEST_REVIEWS,
            variables: {
                skip: 0,
                limit: 20,
            },
        },
        result: {
            data: {
                latestReviews: all_reviews.slice(0, 20),
            },
        },
    },
];

const mockLatestReviewsError = [
    {
        request: {
            query: GET_LATEST_REVIEWS,
            variables: {
                skip: 0,
                limit: 20,
            },
        },
        error: new Error("Error fetching movie"),
    },
];

const renderComponent = (mocks: MockedResponse[] | undefined) => {
    return render(
        <MockedProvider mocks={mocks} addTypename={false}>
            <MemoryRouter>
                <ActivityPage />
            </MemoryRouter>
        </MockedProvider>
    );
};

describe("ActivityPage Snapshots", () => {
    it("matches snapshot when loading", () => {
        const { asFragment } = renderComponent(mockLatestReviewsEmpty);
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot when latest reviews are loaded", async () => {
        const { asFragment } = renderComponent(mockLatestReviews);
        await screen.findByText("Latest Activity");
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot when no reviews are available", async () => {
        const { asFragment } = renderComponent(mockLatestReviewsEmpty);
        await screen.findByText("No reviews have been added yet");
        expect(asFragment()).toMatchSnapshot();
    });

    it("matches snapshot when there is an error", async () => {
        const { asFragment } = renderComponent(mockLatestReviewsError);
        await screen.findByText("Something went wrong!");
        expect(asFragment()).toMatchSnapshot();
    });
});

describe("ActivityPage", () => {
    it("displays 'No reviews have been added yet' on empty reviews", async () => {
        renderComponent(mockLatestReviewsEmpty);
        await waitFor(() =>
            expect(
                screen.getByText("No reviews have been added yet")
            ).toBeInTheDocument()
        );
    });

    it("displays 'Something went wrong!' on error", async () => {
        renderComponent(mockLatestReviewsError);
        await waitFor(() =>
            expect(
                screen.getByText("Something went wrong!")
            ).toBeInTheDocument()
        );
    });

    it("renders 20 reviews", async () => {
        renderComponent(mockLatestReviews);
        //All reviews have username 'test_user'
        await waitFor(() =>
            expect(screen.getAllByText("test_user")).toHaveLength(20)
        );
    });
});
