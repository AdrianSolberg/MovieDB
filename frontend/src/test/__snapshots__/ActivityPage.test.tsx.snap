// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivityPage > matches snapshot when latest reviews are loaded 1`] = `
<DocumentFragment>
  <section
    class="mx-auto mt-8 max-w-6xl px-4 pb-4"
  >
    <h1
      class="mb-6 text-center text-3xl font-bold"
    >
      Latest Activity
    </h1>
    <div>
      <ul
        class="space-y-6"
      >
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
        <li>
          <div
            data-testid="review-card"
          />
        </li>
      </ul>
      <div
        data-testid="loader"
      />
    </div>
  </section>
</DocumentFragment>
`;

exports[`ActivityPage > matches snapshot when loading 1`] = `
<DocumentFragment>
  <section
    class="mt-6 w-dvw text-center"
  >
    <div
      data-testid="loader"
    />
  </section>
</DocumentFragment>
`;

exports[`ActivityPage > matches snapshot when no reviews are available 1`] = `
<DocumentFragment>
  <section
    class="mt-6 text-center"
  >
    <h1
      class="text-2xl"
    >
      No reviews have been added yet
    </h1>
    <a
      class="text-primary hover:underline"
      href="/"
    >
      Return to home page
    </a>
  </section>
</DocumentFragment>
`;

exports[`ActivityPage > matches snapshot when there is an error 1`] = `
<DocumentFragment>
  <section
    class="mt-6 w-dvw text-center"
  >
    <h1
      class="text-2xl"
    >
      Something went wrong!
    </h1>
    <p
      class="text-primary"
    >
      Try to refresh
    </p>
  </section>
</DocumentFragment>
`;
